aries_vcx/src/handlers/out_of_band/receiver/receiver.rs

```diff
@@ -0,0 +1,96 @@
+use crate::handlers::out_of_band::OutOfBand;
+use crate::handlers::connection::connection::Connection;
+use crate::error::prelude::*;
+use crate::messages::a2a::A2AMessage;
+use crate::messages::attachment::AttachmentId;
+use crate::messages::issuance::credential_offer::CredentialOffer;
+use crate::messages::issuance::credential_request::CredentialRequest;
+use crate::messages::issuance::credential::Credential;
+use crate::messages::proof_presentation::presentation_request::PresentationRequest;
+use crate::messages::proof_presentation::presentation::Presentation;
+use crate::messages::connection::invite::{Invitation, PairwiseInvitation};
+use std::convert::TryFrom;
+
+impl OutOfBand {
+    pub fn create_from_a2a_msg(msg: &A2AMessage) -> VcxResult<Self> {
+        match msg {
+            A2AMessage::OutOfBand(oob) => Ok(oob.clone()),
+            _ => Err(VcxError::from(VcxErrorKind::InvalidMessageFormat))
+        }
+    }
+
+    pub fn connection_exists<'a>(&self, connections: Vec<&'a Connection>) -> VcxResult<Option<&'a Connection>> {
```

Comment:
I think before resolving `full_service` of the oob, we should just try to match by DID itself - if available on connection ofcourse.
With current code, if:
- i establish connection via public invite, using their did `sov:12345`
-  institutions updates their `service` on ledger for their did `sov:12345`
- I find scan their OOB message which contains `services: ["sov:12345"]`

the code would determine these connection not to match. Which is not desired behaviour I think. Also, this scenario should be covered by a test

Reply:
Also I've realized I am not quite sure how well the reference return will actually work once we try to expose these features on handle/ffi layer. You'll have to find out the handle of the returned reference, so we could return that handle via FFI. But I guess we'll find out once we start implementing that.

Reply:
Sure this should be done, we discussed this scenario yesterday. There are a few potential problems:
* the `id` field of a service is [not necessarily a did](https://w3c.github.io/did-core/#services)
* the original connection should be updated before reusing it
* we do not distinguish between connections created via public and pairwise did (could be fixed by distinguishing between pw and public dids using a prefix - as we should)

I agree with potential problems incoming regarding the lifetimes. We require the returned connection to live as long as every connection in the vector, which is probably not very practical.

Reply:
> the original connection should be updated before reusing it

Assuming you maintain association `<public_did> : <vcx_connection_object>`, and you bump into the same public_did in some OOB message - you match it, doesn't mean you need to update service of connection's ddo.

> the bootstrap did doc does not contain public did currently and doing so would probably require assuming the id field of a service

That's right, `@id` can be any URI, not necessarily a DID. Quote from the spec you referred to: 
> The value of the id property MUST be a URI conforming to [RFC3986].

So I think need to remember public DID somewhere aside from DDO. I think on our Connection object level.

Reply:
> the original connection should be updated before reusing it

Assuming you maintain association `<public_did> : <vcx_connection_object>`, and you bump into the same public_did in some OOB message - you match it, doesn't mean you need to update service of connection's ddo.

> the bootstrap did doc does not contain public did currently and doing so would probably require assuming the id field of a service

That's right, `@id` can be any URI, not necessarily a DID. Quote from the spec you referred to: 
> The value of the id property MUST be a URI conforming to [RFC3986].

So I think need to remember public DID somewhere aside from DDO. I think on our Connection object level.